### PR TITLE
Add opentelemetry integration to Redix

### PR DESCRIPTION
**Pull request by andrewhr**
## New Project

Initial approach follows Ecto instrumentation, recording spans for all Redix `[:redix, :pipeline, :stop]` events.

The command sanitization is inspired-by and adapted from [Java instrumentation][1], from where I've also copied the actual commands and what configuration should they follow.

Network attributes are tracked via a "sidecar" process, which keeps track of connection attributes also via `telemetry`. This extra bit of bookkeeping is needed as command events doesn't include that piece of information, unfortunately.

[1]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/b2bc41453bec7df15a45a78b29709fdf634e4643/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/db/RedisCommandSanitizer.java

### Checklist

**Acknowledgements**

- [x] I certify that I am authorized to assign ownership of this code to the OpenTelemetry project.
- [x] I acknowledge that ownership of the [Hex](https://hex.pm) package must be given/transferred to the [OpenTelemetry Hex Org](https://hex.pm/orgs/opentelemetry)

**Project Type**

- [x] Instrumentation
- [ ] Exporter
- [ ] Propagator
- [ ] Metapackage
- [ ] Example
- [ ] Util

**Prior Art**

- [ ] Is this an existing project? Includes private libraries being open sourced.
  - [ ] Yes
  - [x] No
- [ ] Are there any other packages on [Hex](https://hex.pm) providing similar functionality?
  
**Existing Project**

- [ ] Moving from an existing public repository? Provide a link in the description.
- [ ] All authors who have contributed code contained in this PR have signed the [CNCF CLA](https://identity.linuxfoundation.org/projects/cncf)

**All Projects**

- [ ] Label created (maintainers)
- [x] [labeler.yml](https://github.com/open-telemetry/opentelemetry-erlang-contrib/blob/main/.github/labeler.yml) updated
- [x] CI
  - [x] Elixir or Erlang language file is updated
  - [x] Test strategy should use the matrix strategy with all supported language/OTP combinations
  - [x] The job is scoped to only run for relevant code changes
- [x] [Codeowners](https://github.com/open-telemetry/opentelemetry-erlang-contrib/blob/main/CODEOWNERS) adds scoped ownership for project maintainers
- [x] License file (Apache2 - copyright to OpenTelemetry Authors)
- [x] All code complies with the [Otel Specification](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification) and Otel Semantic Conventions - [Trace](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) [Metrics](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions)

### CODE REVIEW

File: instrumentation/opentelemetry_redix/README.md
Comment (andrewhr) [reply instrumentation/opentelemetry_redix/README.md:9]: Sorry for the delay. I've took another try (explicitly mentioning Hammer as suggested).

What do you think @tsloughter?

**Comment by bryannaegele** [comment]
@andrewhr I updated your initial comment with the new project checklist. I don't know of a fancy way to make it selectable like issue templates. 😕

**Comment by andrewhr** [comment]
Thanks @bryannaegele, and sorry for the delay. I think we should be good - the missing piece was CODEOWNERS. This change *should be compliant*, as I've worked with the specs open at my side 😅 Also, the naming/command parsing was adapted from another compliant implementation.

Having said that, it's a best effort and some attribute may be missing - another pair of eyes may be helpful in that regard.

File: instrumentation/opentelemetry_redix/README.md
Comment (tsloughter) [reply instrumentation/opentelemetry_redix/README.md:9]: Sorry for the delay, this looks good.

File: CODEOWNERS
Comment (tsloughter) [inline CODEOWNERS:20]: @bryannaegele  Aren't the submitters supposed to be codeowners instead of us?

Comment (andrewhr) [reply CODEOWNERS]: That's wasn't totally clear to me. Updated to my user, following the pattern I saw at #6

File: instrumentation/opentelemetry_redix/README.md
Comment (andrewhr) [reply instrumentation/opentelemetry_redix/README.md:9]: Thank you, I appreciate your feedback!

**Review by bryannaegele — APPROVED** [review]
I don't have any real experience with Redix but everything looks good. 👍🏻

File: instrumentation/opentelemetry_redix/test/opentelemetry_redix_test.exs
Comment (bryannaegele) [inline instrumentation/opentelemetry_redix/test/opentelemetry_redix_test.exs:82]: The failing test appears to be that this is emitted with new lines between commands.

**Review by bryannaegele — COMMENTED** [review]
There's a bunch of warnings in the test about failing to export spans to `4317` but I don't see where that's originating from.

File: instrumentation/opentelemetry_redix/test/opentelemetry_redix_test.exs
Comment (andrewhr) [reply instrumentation/opentelemetry_redix/test/opentelemetry_redix_test.exs:82]: 🙈 updated the implementation after looking at some results, forgot to update the tests (and check here). Sorry for that!

**Comment by andrewhr** [comment]
About the export warnings, AFAICT that comes from `otel_batch_processor` which is set on test environment (I assume for the test helpers to work properly). 

As soon `otel_batch_processor` starts it will then proceed to setup the `otel_exporter` which by default exports to OTEL collector default port. The setup block races against that but eventually some warnings get reported.

I was able to reproduce the same issue locally with Phoenix and Ecto instrumentation (Ecto reports are bit different because Elixir logger isn't started there).

I got a fix here in this PR, and will prepare some for them as well (in a separate PR).

**Comment by andrewhr** [comment]
Rebased on latest `main`, to fix merge conflicts.

**Comment by andrewhr** [comment]
Hey folks, updated to `1.0.0-rc.4` (I know, 1.0 is around the corner).

Worth noting, with `:simple_span_processor` I was able to simplify a bit the test setup. Please let me know what do you think (and if it's better to wait for SDK 1.0)

**Comment by tsloughter** [comment]
Thanks @andrewhr . It is best to wait for 1.0, I'm in the process of releasing it to hex as I write this.

**Comment by andrewhr** [comment]
@tsloughter updated to `1.0.0`, thank you for all the work on that release! 🙌

**Review by tsloughter — APPROVED** [review]

**Comment by bryannaegele** [comment]
Hey @andrewhr - there were some changes recently to the CI files which don't appear to have been updated in this PR. Can you open a new PR with those? https://github.com/open-telemetry/opentelemetry-erlang-contrib/pull/55/files#diff-567a3fdf49d6f0f0119dda6d1c098fcb5ee3785ee956481b3daaf5a2f246076f

**Comment by andrewhr** [comment]
Sure @bryannaegele, my bad I miss that update. Here goes https://github.com/open-telemetry/opentelemetry-erlang-contrib/pull/58

**Pull request merged**
